src/base/TypeUtil.ml

```diff
@@ -147,33 +152,50 @@ functor
         tvars : (string, R.rep) Hashtbl.t;
       }
 
+      type restore =
+        (* This relies on Caml.Hashtbl allowing adding
+         * bindings on top of existing ones and removing them
+         * to restore the older binding. Restoring working in
+         * FIFO order, i.e., most recent change restored first. *)
+        | AddT of string * resolve_result
+        | RemT of string
+        (* | AddV of string * R.rep *)
+        | RemV of string
+
```

Comment:
If you go with the `restore` rather than `restore list` suggestion, then you will need something like

```
type restore_op =
  | AddT of ...
  etc.

type restore = restore_op list

type additional_restore r rs = r :: rs
```

The result will be that the from the point of view of the user of `TEnv`, there will just be one `restore` object for whichever operation they have performed, and there is no way to inspect or manipulate them (at the moment the restoration can be manipulated by removing `restore` objects from the returned lists).